2-advanced-topics/data-pipelines/intro_data_pipelines.ipynb

```diff
@@ -0,0 +1,1037 @@
+     "showTitle": false,
+     "title": ""
+    }
+   },
+   "source": [
+    "## Gold\n",
+    "\n",
+    "In the gold stage of the pipeline (gold.py in the project folder), we finalize the dataset for analysis. The focus here is on filtering, selecting key columns, and ensuring the data is structured for consumption.\n",
+    "\n",
+    "This stage refines the data into a ready-to-use form. We retain columns deemed necessary for the analysis, as the remaining columns are available in the silver table if required later."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 0,
+   "metadata": {
+    "application/vnd.databricks.v1+cell": {
+     "cellMetadata": {},
+     "inputWidgets": {},
+     "nuid": "3f03058a-35da-43b5-a8c6-c9bd0ac209c8",
+     "showTitle": false,
+     "title": ""
+    }
+   },
+   "outputs": [],
+   "source": [
+    "gold_df = (spark.read.table(f\"{database_name}.kenya_silver\")\n",
+    "    .filter(col('year') != 2015) # Exclude 2015 data\n",
+    "    .withColumn('country_name', lit('Kenya'))\n",
+    "    .select('country_name',\n",
+    "            'year',\n",
+    "            col('Initial_Budget_Printed_Estimate').alias('approved').cast(DoubleType()),\n",
+    "            col('Final_Budget_Approved_Estimate').alias('revised'),\n",
+    "            col('`Final_Expenditure_Total_Payment_Comm.`').alias('executed'),\n",
+    "            'admin0',\n",
+    "            'admin1',\n",
+    "            'admin2',\n",
+    "            'geo1',\n",
+    "            'is_foreign',\n",
+    "            'func_sub'\n",
+    "           )\n",
+    ")\n",
+    "\n",
+    "# Save to gold table\n",
+    "gold_df.write.format(\"delta\").mode(\"overwrite\").saveAsTable(f\"{database_name}.kenya_gold\")\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "application/vnd.databricks.v1+cell": {
+     "cellMetadata": {},
+     "inputWidgets": {},
+     "nuid": "2b285fd1-0002-4ed8-8736-26165242f9c9",
+     "showTitle": false,
+     "title": ""
+    }
+   },
+   "source": [
+    "## Aggregation and appending additional columns\n",
+    "\n",
+    "In the above steps we cleaned the original source BOOST Kenya data through multiple stages. In addition to this, we might need external datasources to append to this data. In our example, we want to add information about per capita spending.\n",
+    "\n",
+    "In this case, we want to get the subnational population from an external source (an API) and then aggregate our kenya_gold data to get a table containing the spending by *geo1* region and year. Since we referenced this very dataset [above](#do-we-already use-data-pipelines?-or-are-they-completely-new?), we will not go over the code again but it can be accessed [here](subnational_population.ipynb). Then we will append a new column called per_capita_spending to this aggregated table. We finally save this table so that these descriptive figures are readily available. "
```

Comment:
```suggestion
    "In this case, we want to get the subnational population from an external source (an API) and then aggregate our kenya_gold data to get a table containing the spending by *geo1* region and year. Since we referenced this very dataset [above](#Do-we-already-use-data-pipelines?-Or-are-they-completely-new?), we will not go over the code again but it can be accessed [here](subnational_population.ipynb). Then we will append a new column called per_capita_spending to this aggregated table. We finally save this table so that these descriptive figures are readily available. "
```